File: ga/4omini_result_2024_7/result/queue_l_mini/20_queue_l.c
use std::io;
use std::ptr;
use std::alloc::{alloc, dealloc, Layout};
use std::mem;

const MIN_QUEUE_SIZE: usize = 5;

#[derive(Debug)]
struct Queue {
    capacity: usize,
    front: usize,
    rear: usize,
    size: usize,
    array: *mut i32, // Using raw pointer for dynamic array allocation
}

impl Queue {
    fn new(max_elements: usize) -> Result<Self, &'static str> {
        if max_elements < MIN_QUEUE_SIZE {
            return Err("Queue size is too small");
        }

        let layout = Layout::from_size_align(max_elements * mem::size_of::<i32>(), mem::align_of::<i32>())
            .expect("Failed to create layout");
        
        unsafe {
            let array = alloc(layout);
            if array.is_null() {
                return Err("Out of space!!!");
            }

            Ok(Queue {
                capacity: max_elements,
                front: 0,
                rear: (max_elements - 1),
                size: 0,
                array,
            })
        }
    }

    fn is_empty(&self) -> bool {
        self.size == 0
    }

    fn is_full(&self) -> bool {
        self.size == self.capacity
    }

    fn make_empty(&mut self) {
        self.size = 0;
        self.front = 0;
        self.rear = self.capacity - 1;
    }

    fn enqueue(&mut self, x: i32) -> Result<(), &'static str> {
        if self.is_full() {
            return Err("Full queue");
        } else {
            self.size += 1;
            self.rear = (self.rear + 1) % self.capacity;
            unsafe {
                *(self.array.offset(self.rear as isize)) = x;
            }
            Ok(())
        }
    }

    fn front(&self) -> Result<i32, &'static str> {
        if self.is_empty() {
            return Err("Empty queue");
        }
        unsafe {
            Ok(*(self.array.offset(self.front as isize)))
        }
    }

    fn dequeue(&mut self) -> Result<(), &'static str> {
        if self.is_empty() {
            return Err("Empty queue");
        } else {
            self.size -= 1;
            self.front = (self.front + 1) % self.capacity;
            Ok(())
        }
    }

    fn front_and_dequeue(&mut self) -> Result<i32, &'static str> {
        let value = self.front()?;
        self.dequeue()?;
        Ok(value)
    }
}

impl Drop for Queue {
    fn drop(&mut self) {
        if !self.array.is_null() {
            unsafe {
                let layout = Layout::from_size_align(self.capacity * mem::size_of::<i32>(), mem::align_of::<i32>())
                    .expect("Failed to create layout");
                dealloc(self.array, layout);
            }
        }
    }
}

fn main() {
    let mut input = String::new();
    println!("Enter queue length:");
    io::stdin().read_line(&mut input).expect("Failed to read line");
    let queue_length: usize = input.trim().parse().expect("Invalid input or queue size is too small");

    let mut queue = Queue::new(queue_length).expect("Failed to create queue");

    // Test: Attempt to dequeue from an empty queue
    if !queue.is_empty() {
        queue.dequeue().expect("Failed to dequeue");
    } else {
        println!("Queue is empty, cannot dequeue.");
    }

    // Enqueue elements and check if the queue is full
    for i in 0..(queue_length - 1) {
        if queue.is_full() {
            println!("Queue is full");
        } else {
            queue.enqueue(i).expect("Failed to enqueue");
        }
    }

    // Test: Attempt to enqueue to a full queue
    if queue.is_full() {
        println!("Queue is full, cannot enqueue.");
    } else {
        println!("Enter a value to enqueue:");
        let mut value_input = String::new();
        io::stdin().read_line(&mut value_input).expect("Failed to read line");
        let enqueue_value: i32 = value_input.trim().parse().expect("Invalid input");
        queue.enqueue(enqueue_value).expect("Failed to enqueue");
    }

    // Use front_and_dequeue to get and remove elements
    while !queue.is_empty() {
        match queue.front_and_dequeue() {
            Ok(value) => println!("FrontAndDequeue: {}", value),
            Err(e) => println!("{}", e),
        }
    }

    // Enqueue elements again
    for i in 0..(queue_length - 2) {
        queue.enqueue(i).expect("Failed to enqueue");
    }

    // Use front and dequeue
    while !queue.is_empty() {
        match queue.front() {
            Ok(value) => {
                println!("Front: {}", value);
                queue.dequeue().expect("Failed to dequeue");
            }
            Err(e) => println!("{}", e),
        }
    }
}

File: ga/glm4_result_2024_6/result/queue_l_glm/20_queue_l.rs
use std::io;
use std::io::{Error, ErrorKind};

const MIN_QUEUE_SIZE: usize = 5;

#[derive(Debug)]
struct Queue {
    capacity: usize,
    front: usize,
    rear: usize,
    size: usize,
    array: Vec<i32>,
}

impl Queue {
    fn new(capacity: usize) -> Self {
        if capacity < MIN_QUEUE_SIZE {
            panic!("Queue size is too small");
        }
        Queue {
            capacity,
            front: 0,
            rear: 0,
            size: 0,
            array: vec![0; capacity],
        }
    }

    fn is_empty(&self) -> bool {
        self.size == 0
    }

    fn is_full(&self) -> bool {
        self.size == self.capacity
    }

    fn enqueue(&mut self, value: i32) {
        if self.is_full() {
            panic!("Queue is full");
        }
        self.size += 1;
        self.array[self.rear] = value;
        self.rear = (self.rear + 1) % self.capacity;
    }

    fn front(&self) -> i32 {
        if self.is_empty() {
            panic!("Queue is empty");
        }
        self.array[self.front]
    }

    fn dequeue(&mut self) {
        if self.is_empty() {
            panic!("Queue is empty");
        }
        self.size -= 1;
        self.front = (self.front + 1) % self.capacity;
    }
}

fn main() {
    let mut queue_size = 0;
    println!("Enter the queue size: ");
    io::stdin()
        .read_line(&mut queue_size)
        .expect("Failed to read line");

    let queue_size: usize = queue_size.trim().parse().expect("Invalid input or queue size is too small");

    let mut queue = Queue::new(queue_size);

    // Test: Attempt to dequeue from an empty queue
    if queue.is_empty() {
        println!("Queue is empty, cannot dequeue.");
    } else {
        queue.dequeue();
    }

    // Enqueue elements and check if the queue is full
    for i in 0..queue_size - 1 {
        if queue.is_full() {
            println!("Queue is full");
        } else {
            queue.enqueue(i);
        }
    }

    // Test: Attempt to enqueue to a full queue
    let mut enqueue_value = String::new();
    println!("Enter a value to enqueue: ");
    io::stdin()
        .read_line(&mut enqueue_value)
        .expect("Failed to read line");
    let enqueue_value: i32 = enqueue_value.trim().parse().expect("Invalid input");

    if queue.is_full() {
        println!("Queue is full, cannot enqueue.");
    } else {
        queue.enqueue(enqueue_value);
    }

    // Use FrontAndDequeue to get and remove elements
    while !queue.is_empty() {
        println!("FrontAndDequeue: {}", queue.front());
        queue.dequeue();
    }

    // Enqueue elements again
    for i in 0..queue_size - 2 {
        queue.enqueue(i);
    }

    // Use Front and Dequeue
    while !queue.is_empty() {
        println!("Front: {}", queue.front());
        queue.dequeue();
    }
}